File: SiderShopping/app/src/main/java/com/mmart/sidershopping/view/productdetail/ProductDetailActivity.java
package com.mmart.sidershopping.view.productdetail;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.text.Html;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.mmart.sidershopping.R;
import com.mmart.sidershopping.model.entity.Product;
import com.squareup.picasso.Picasso;

import java.text.DecimalFormat;

public class ProductDetailActivity extends AppCompatActivity {

    private static DecimalFormat df2 = new DecimalFormat("#.##");

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_product_detail);

        setActionBar();

        final int key = getKey();

        setProduct(key);
    }

    private void setProduct(final int key) {
        final ProductDetailViewModel productDetailViewModel = ViewModelProviders.of(this).get(ProductDetailViewModel.class);

        productDetailViewModel.getProduct(key).observe(this, new Observer<Product>() {
            @Override
            public void onChanged(@Nullable Product product) {

                ImageView image = findViewById(R.id.product_image);
                TextView name = findViewById(R.id.product_name);
                TextView oldPrice = findViewById(R.id.product_old_price);
                TextView newPrice = findViewById(R.id.product_new_price);
                TextView description = findViewById(R.id.product_description);

                Picasso.get()
                        .load(product.getUrlImagem())
                        .fit()
                        .centerCrop()
                        .into(image);

                name.setText(product.getNome());
                description.setText(Html.fromHtml(product.getDescricao()));
                oldPrice.setText("De: " + df2.format(product.getPrecoDe()));
                newPrice.setText("Por " + df2.format(product.getPrecoPor()));
            }
        });

        FloatingActionButton button = findViewById(R.id.floating_button);

        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                reserve(key, productDetailViewModel);

                finish();
            }
        });
    }

    private void reserve(int key, ProductDetailViewModel productDetailViewModel) {
        productDetailViewModel.reserve(key).observe(this, new Observer<String>() {
            @Override
            public void onChanged(@Nullable String message) {
                Toast.makeText(ProductDetailActivity.this, message, Toast.LENGTH_SHORT).show();
            }
        });
    }

    private int getKey() {
        Intent intent = getIntent();
        return intent.getIntExtra("key", 0);
    }

    private void setActionBar() {
        ActionBar ab = getSupportActionBar();
        ab.setDisplayHomeAsUpEnabled(true);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }
}

File: app/src/main/java/br/com/amedigital/product/ProductDetailActivity.kt
package br.com.amedigital.product

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.design.widget.AppBarLayout
import android.support.v7.app.AlertDialog
import android.text.Html
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import br.com.amedigital.R
import br.com.amedigital.model.Product
import br.com.amedigital.network.LojinhaServiceEndPoint
import br.com.amedigital.util.formatToMoney
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.activity_product_detail.*
import kotlinx.android.synthetic.main.partial_product_detail.*

class ProductDetailActivity : AppCompatActivity(), ProductContract.ViewReserveProduct {

    companion object {
        const val PRODUCT = "product"
    }

    var product : Product? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_product_detail)
        getDataExtra(intent.extras)
        setCollapsingButtonIcon()
        buttonSaveProduct()
    }

    private fun getDataExtra(extras: Bundle?) {
        product = extras?.getSerializable(PRODUCT) as Product
        product?.apply {
            setToolbar()
            bindProduct(this)
        }
    }

    private fun setToolbar(){
        toolbar.title = ""
        toolbar.setNavigationIcon(R.drawable.ic_arrow_back_black_24px)
        setSupportActionBar(toolbar)
    }

    private fun setCollapsingButtonIcon() {
        app_bar.addOnOffsetChangedListener(object : AppBarLayout.OnOffsetChangedListener {
            var isShow = false
            var scrollRange = -1

            override fun onOffsetChanged(appBarLayout: AppBarLayout, verticalOffset: Int) {
                if (scrollRange == -1) {
                    scrollRange = appBarLayout.totalScrollRange
                }
                if (scrollRange + verticalOffset == 0) {
                    isShow = true
                    toolbar.setNavigationIcon(R.drawable.ic_arrow_back_white_24px)
                } else if (isShow) {
                    isShow = false
                    toolbar.setNavigationIcon(R.drawable.ic_arrow_back_black_24px)
                }
            }
        })
    }

    private fun reserveProduct() {
        product?.apply {
            val call = LojinhaServiceEndPoint().setReservation(id)
            val presenter = ProductPresenter()
            presenter.reserve(call, this@ProductDetailActivity)
        }
    }

    private fun bindProduct(product: Product) {
        Picasso.get()
            .load(product.urlImage)
            .into(imageViewProductDesc)
        textViewCategoryName.text = product.category?.description
        textViewDe.text = "De ${product.priceOf.formatToMoney()}"
        textViewPor.text = "Por ${product.priceFor.formatToMoney()}"
        textViewProdName.text = product.name
        textViewDescProd.text = Html.fromHtml(product.description)
    }

    private fun buttonSaveProduct() {
        floatingButtonReserve.setOnClickListener {
            floatingButtonReserve.isEnabled = false
            reserveProduct()
        }
    }

    override fun setProgressReserve(active: Boolean) {
        floatingButtonReserve.isEnabled = !active
        progressBarReserve.visibility = if (active) View.VISIBLE else View.GONE
    }

    override fun successReserve() {
        val builder = AlertDialog.Builder(this)
            builder.setCancelable(false)
            builder.setMessage(getString(R.string.success_reserve))
            builder.setPositiveButton(getString(R.string.ok)){ dialog, which ->
                dialog.dismiss()
                finish()
            }
            builder.create().show()
    }

    override fun showErrorReserve(message: String) {
        Toast.makeText(this@ProductDetailActivity, message, Toast.LENGTH_LONG).show()
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item?.itemId == android.R.id.home) {
            finish()
        }
        return super.onOptionsItemSelected(item)
    }
}

